apis/core.oam.dev/v1alpha2/applicationrevision_types.go

```diff
@@ -40,11 +41,23 @@ type ApplicationRevisionSpec struct {
 	ScopeDefinitions map[string]ScopeDefinition `json:"scopeDefinitions,omitempty"`
 
 	// Components records the rendered components from Application, it will contains the whole K8s CR of workload in it.
-	Components []*Component `json:"components"`
+	Components map[string]RawExtensionWrapper `json:"components,omitempty"`
 
 	// ApplicationConfiguration records the rendered applicationConfiguration from Application,
 	// it will contains the whole K8s CR of trait and the reference component in it.
-	ApplicationConfiguration ApplicationConfiguration `json:"applicationConfiguration"`
+	// +kubebuilder:validation:EmbeddedResource
+	// +kubebuilder:pruning:PreserveUnknownFields
+	ApplicationConfiguration runtime.RawExtension `json:"applicationConfiguration"`
```

Comment:
we are ok with ApplicationConfiguration without any meta field

Reply:
AppDeploy controller will need the labels etc.